ctf_sec

medium

# one plugin's withdraw and deposit failure blocks the rest of the plugin's withdraw and deposit and revert the transaction.

## Summary

one plugin's withdraw and deposit failure blocks the rest of the plugin's withdraw and deposit and revert the transaction.

## Vulnerability Detail

currently, the deposit function would deposit the token into the plugin to generate yield, however, the deposit function iterate over the plugin count and consume the deposit limit. If one deposit revert, the whole transaction revert.

Consider the case below,

there are 3 plugins added in the Vault.sol

the first plugin is broken, and cannot process deposit,

the second plugin can take half of the deposit with sufficient availableForDeposit limit

the third plugin can take half of the deposit with sufficient availableForDeposit limit.

the user call deposit, the transaction is supposed to be processed because although the first plugin failures, the second and thrid plugin can handle the deposit.

but under the current implementation the transaction revert when the first plugin's deposit method is called.

https://github.com/sherlock-audit/2022-10-mycelium/blob/main/mylink-contracts/src/Vault.sol#L460-L470

Same issue is also presented in _ensureLinkAmount

https://github.com/sherlock-audit/2022-10-mycelium/blob/main/mylink-contracts/src/Vault.sol#L502-L513

suppose we have three plugin

the first plugin fails to process the withdraw,

but the second and third plugin can process the withdraw,

the transaction revert when the first withdraw is called.

## Impact

Transaction is reverted although the underlying plugin is capable of processing the withdraw and deposit request.

## Code Snippet



## Tool used

Manual Review

## Recommendation

We recommend add error handling when doing deposit and withdraw with plugin, and adding a fallback implementation:

when first plugin fails, go to the next one.

I think it is like a web interface with three server. 

We have three server,  if one primary server is down, the work should be fall-backed to second and third server instead of not processing the user request.
